### PR TITLE
[Google Blockly] Bitmap field and new Sprite Lab blocks

### PR DESCRIPTION
**Pull request by mikeharv**
This branch adds a new Blockly plugin [@blockly/field-bitmap](https://www.npmjs.com/package/@blockly/field-bitmap) which allows for user-inputted pixel grids. This new field class is extended and customized in our usual fashion and hooked up as a custom input type for Sprite Lab blocks.  Three new blocks are created which allow for users to create a bitmap representation of a layout for sprites. These blocks are intended for use in the new game development module and build off of the work in https://github.com/code-dot-org/code-dot-org/pull/57022.

image
image

The two following animations were recorded before the "environment" label was converted to a dropdown field (seen above). However, there's no other difference in appearance. Standard block with built-in field:

https://github.com/code-dot-org/code-dot-org/assets/43474485/57b682a0-702e-4a5e-ab09-3a24f9314485

Block with separate Array input and shadow block:

Note that default size grid is 8x8. This was chosen as it aligns well to Sprite Lab's 400x400 canvas in that sprites are automatically scaled to 50%. Other grid sizes can be specified in the block XML (for start blocks or toolboxes), but students cannot currently adjust the size of the grid themselves. New sprites are scaled and positioned automatically regardless of grid size.

## Extended field class
As stated in the plugin's documentation, it does support serialization to XML. Fortunately, we can accomplish this easily by using JSON methods to stringify and parse the field value. The `toXml` and `fromXml` hooks work with the old XML serialization system and have been added to our class.

## Scale and position calculations
The scale is determined based on the app height (400) and the number of rows in the array. For example, with a 4x4 grid, the scale is 100 and the sprites are rendered with a size of 100x100. Sprite x/y coordinates represent the center of the sprite. So, to position the sprites we offset x and y by _half_ of the scale.
imageimage

As like the previous work in this space (https://github.com/code-dot-org/code-dot-org/pull/57022), the exact block configuration and helper code will live on levelbuilder and so we should expect some iteration outside of the normal PR flow between now and piloting.

## Follow-up work

Further customize the field. This work is also up for review in a branch based off of this one:
- https://github.com/code-dot-org/code-dot-org/pull/57107

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/cdoFieldBitmap.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * Custom FieldBitmap class with additional hooks for XML serialization.
+ */
+export class CdoFieldBitmap extends FieldBitmap {
+  /**
+   * Constructs a new instance of CdoFieldBitmap.
+   * @param {number[][] | null} value - The initial value of the field, represented as a 2D array of any length, or null/undefined.
+   * @param {object | null} options - The options for the field, can be an object or null/undefined.
+   * @param {object | null} config - Additional configuration options, can be an object or null/undefined.
+   */
+  constructor(value?: number[][], options?: object, config?: object) {
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoFieldBitmap.ts]: if you want to be able to pass in null the types here should be `number[][] | null`, etc. Adding `?` allows undefined but not null.

File: apps/src/p5lab/spritelab/commands/spriteCommands.js [hunk 2]
```diff
@@ -82,6 +83,19 @@ export const commands = {
     }
   },
 
+  makeEnvironmentSprites(animation, group, bitmap) {
+    const scale = APP_HEIGHT / bitmap.length;
+    for (let i = 0; i < bitmap.length; i++) {
+      for (let j = 0; j < bitmap[0].length; j++) {
+        const location = {
+          x: scale / 2 + scale * i,
```
Comment (molly-moen) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: mind adding a comment documenting this calculation?

**Review by molly-moen — APPROVED** [review]
Looks good. One question to think about as we iterate: can we make the block itself smaller? It makes the block it is connected to quite tall and will mean more scrolling for students on small screens. **Edit** I see you did that in the follow up PR, never mind!

**Comment by mikeharv** [comment]
This branch has been updated to include all changes from the PR linked below. Please see that page for additional details in its description.
- https://github.com/code-dot-org/code-dot-org/pull/57107

File: apps/src/blockly/addons/cdoFieldBitmap.ts (hunk 1 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoFieldBitmap.ts]: Updated here: https://github.com/code-dot-org/code-dot-org/pull/57107

File: apps/src/p5lab/spritelab/commands/spriteCommands.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: Updated here: https://github.com/code-dot-org/code-dot-org/pull/57107

**Pull request merged**